### PR TITLE
Add dependency filter

### PR DESCRIPTION
**Pull request by basilfx**
# Pull Request

## Summary
An additional regex filter can be supplied, to only include (or exclude) certain dependencies in the resulting output. This applies to collated dependencies as well.

My use case: only 'complain' about transitive dependencies from `My.Company.*`, or ignore `Popular.Dependency`.

### Previous
No filtering possible.

### New
Filtering is possible.

## Breaking Changes
The API has changed in a compatible way.

```diff
-NuGetTransitiveDependencyFinder.ITransitiveDependencyFinder.Run(string? projectOrSolutionPath, bool collateAllDependencies) -> NuGetTransitiveDependencyFinder.Output.Projects!
+NuGetTransitiveDependencyFinder.ITransitiveDependencyFinder.Run(string? projectOrSolutionPath, bool collateAllDependencies, System.Text.RegularExpressions.Regex? filter) -> NuGetTransitiveDependencyFinder.Output.Projects!
```

### CODE REVIEW

File: src/Product/NuGetTransitiveDependencyFinder.ConsoleApp/Input/ICommandLineOptions.cs [hunk 1]
```diff
@@ -20,5 +20,10 @@ internal interface ICommandLineOptions
         /// Gets the file name of the .NET project or solution to analyze.
         /// </summary>
         public string? ProjectOrSolution { get; init; }
+
+        /// <summary>
+        /// Gets the file name of the .NET project or solution to analyze.
```
[the remark this answers is not in the document]
Comment (basilfx) [reply src/Product/NuGetTransitiveDependencyFinder.ConsoleApp/Input/ICommandLineOptions.cs]: Fixed

File: src/Product/NuGetTransitiveDependencyFinder.ConsoleApp/Input/CommandLineOptions.cs [hunk 2]
```diff
@@ -25,5 +25,13 @@ internal class CommandLineOptions : ICommandLineOptions
...
             ResourceType = typeof(CommandLineHelp))]
         public string? ProjectOrSolution { get; init; }
+
+        /// <inheritdoc/>
+        [Option(
+            'f',
+            "filter",
+            HelpText = nameof(Filter),
+            ResourceType = typeof(CommandLineHelp))]
+        public string? Filter { get; init; }
```
[the remark this answers is not in the document]
Comment (basilfx) [reply src/Product/NuGetTransitiveDependencyFinder.ConsoleApp/Input/CommandLineOptions.cs]: It seems that it is supported out of the box, although I cannot add RegexOptions.Compiled.

File: src/Product/NuGetTransitiveDependencyFinder/ITransitiveDependencyFinder.cs [hunk 3]
```diff
@@ -20,9 +20,14 @@ public interface ITransitiveDependencyFinder : IDisposable
...
+        /// <param name="filter">An optional regular expression pattern, to match certain dependencies. It will filter
+        /// non-transitive dependencies as well, if <paramref name="collateAllDependencies"/> is
+        /// <see langword="true"/>.</param>
         /// <returns>The transitive NuGet dependency information, which can be processed for display.</returns>
         /// <exception cref="ArgumentNullException"><paramref name="projectOrSolutionPath"/> is
         /// <see langword="null"/>.</exception>
-        public Projects Run(string? projectOrSolutionPath, bool collateAllDependencies);
+        /// <exception cref="ArgumentException"><paramref name="filter"/> is not a valid regular
+        /// expression.</exception>
+        public Projects Run(string? projectOrSolutionPath, bool collateAllDependencies, string? filter);
```
[the remark this answers is not in the document]
Comment (basilfx) [reply src/Product/NuGetTransitiveDependencyFinder/ITransitiveDependencyFinder.cs]: Fixed this.

**Comment by basilfx** [comment]
Updated and rebased the PR. Added the changes as separate commits, which I can squash before merging.

File: src/Product/NuGetTransitiveDependencyFinder.ConsoleApp/Input/CommandLineOptions.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (muiriswoulfe) [reply src/Product/NuGetTransitiveDependencyFinder.ConsoleApp/Input/CommandLineOptions.cs]: Thanks for checking

**Review by muiriswoulfe — APPROVED** [review]

**Comment by basilfx** [comment]
Squashed and signed the commits.

**Comment by basilfx** [comment]
@muiriswoulfe Anything I need to change? I will update the other PRs once this is in :-)

**Comment by muiriswoulfe** [comment]
> @muiriswoulfe Anything I need to change? I will update the other PRs once this is in :-)

No, everything looks good. I'll squash and merge it now. Had just missed the last update 🙂

**Pull request merged**